apps/forms/form-dialog-alert/src/app/pages/join.component.ts

```diff
@@ -1,16 +1,40 @@
-import { ChangeDetectionStrategy, Component } from '@angular/core';
+import {
+  ChangeDetectionStrategy,
+  Component,
+  inject,
+  viewChild,
+} from '@angular/core';
+import { MatDialog } from '@angular/material/dialog';
+import { Observable, of } from 'rxjs';
+import { ComponentCanDeactivate } from '../guards/pending-changes.guard';
+import { UnloadGuardDirective } from '../guards/unload-window.directive';
+import { AlertDialogComponent } from '../ui/dialog.component';
 import { FormComponent } from '../ui/form.component';
 
 @Component({
   standalone: true,
-  imports: [FormComponent],
+  imports: [FormComponent, UnloadGuardDirective],
   template: `
     <section class="mx-auto	max-w-screen-sm">
       <div class="rounded-lg bg-white p-8 shadow-lg lg:p-12">
-        <app-form />
+        <app-form appUnloadGuard />
       </div>
     </section>
   `,
   changeDetection: ChangeDetectionStrategy.OnPush,
 })
-export class JoinComponent {}
+export class JoinComponent implements ComponentCanDeactivate {
+  private dialog = inject(MatDialog);
+  formComponent = viewChild(FormComponent);
+
+  canDeactivate(): boolean | Observable<boolean> {
+    if (this.formComponent()!.form.dirty) {
```

Comment:
This gives a `Forbidden non null assertion` eslint error.  You can change it to `this.formComponent()?.form.dirty`.